src/host/outputStream.cpp

```diff
@@ -608,54 +593,61 @@ bool ConhostInternalGetSet::MoveToBottom() const
+// - the COLORREF value for the color at that index in the table.
+COLORREF ConhostInternalGetSet::GetColorTableEntry(const size_t tableIndex) const noexcept
+try
 {
-    return SUCCEEDED(DoSrvPrivateGetColorTableEntry(index, value));
-}
+    auto& g = ServiceLocator::LocateGlobals();
+    auto& gci = g.getConsoleInformation();
 
-// Method Description:
-// - Connects the PrivateSetColorTableEntry call directly into our Driver Message servicing
-//      call inside Conhost.exe
-// Arguments:
-// - index: the index in the table to change.
-// - value: the new RGB value to use for that index in the color table.
-// Return Value:
-// - true if successful (see DoSrvPrivateSetColorTableEntry). false otherwise.
-bool ConhostInternalGetSet::PrivateSetColorTableEntry(const size_t index, const COLORREF value) const noexcept
-{
-    return SUCCEEDED(DoSrvPrivateSetColorTableEntry(index, value));
+    return gci.GetColorTableEntry(tableIndex);
 }
-
-// Method Description:
-// - Connects the PrivateSetDefaultForeground call directly into our Driver Message servicing
-//      call inside Conhost.exe
-// Arguments:
-// - value: the new RGB value to use, as a COLORREF, format 0x00BBGGRR.
-// Return Value:
-// - true if successful (see DoSrvPrivateSetDefaultForegroundColor). false otherwise.
-bool ConhostInternalGetSet::PrivateSetDefaultForeground(const COLORREF value) const noexcept
+catch (...)
 {
-    return SUCCEEDED(DoSrvPrivateSetDefaultForegroundColor(value));
+    return INVALID_COLOR;
 }
 
 // Method Description:
-// - Connects the PrivateSetDefaultBackground call directly into our Driver Message servicing
-//      call inside Conhost.exe
+// - Updates the value in the colortable at index tableIndex to the new color
+//   color. color is a COLORREF, format 0x00BBGGRR.
 // Arguments:
-// - value: the new RGB value to use, as a COLORREF, format 0x00BBGGRR.
+// - tableIndex: the index of the color table to update.
+// - color: the new COLORREF to use as that color table value.
 // Return Value:
-// - true if successful (see DoSrvPrivateSetDefaultBackgroundColor). false otherwise.
-bool ConhostInternalGetSet::PrivateSetDefaultBackground(const COLORREF value) const noexcept
+// - true if successful. false otherwise.
+bool ConhostInternalGetSet::SetColorTableEntry(const size_t tableIndex, const COLORREF color) noexcept
+try
 {
-    return SUCCEEDED(DoSrvPrivateSetDefaultBackgroundColor(value));
+    auto& g = ServiceLocator::LocateGlobals();
+    auto& gci = g.getConsoleInformation();
+
+    gci.SetColorTableEntry(tableIndex, color);
+
+    // If we're setting the default foreground or background colors
+    // we need to make sure the index is correctly set as well.
+    if (tableIndex == TextColor::DEFAULT_FOREGROUND)
+    {
+        gci.SetDefaultForegroundIndex(TextColor::DEFAULT_FOREGROUND);
+    }
+    if (tableIndex == TextColor::DEFAULT_BACKGROUND)
+    {
+        gci.SetDefaultBackgroundIndex(TextColor::DEFAULT_BACKGROUND);
+    }

```

Comment:
At some point I'm hoping this code might be moved up to the `AdaptDispatch`/`TerminalDispatch` level, since there are separate code paths for the default colors, which should be able to set these index values unconditionally. I just didn't want to overcomplicate the PR with the introduction of index APIs until we really need them.

Reply:
Yeah I feel you there. There's going to be a lot more we want to share at that level going forward.